### PR TITLE
Blog Post - Karate Test Automation

### PR DESCRIPTION
**Pull request by sdewar-scottlogic**
Please add a direct link to your post here:

https://sdewar-scottlogic.github.io/blog/2024/10/28/an-introduction-into-karate-test-automation.html

Have you (please tick each box to show completion):

 - [x] Added your blog post to a single category?
 - [x] Added a brief summary for your post? Summaries should be roughly two sentences in length and give potential readers a good idea of the contents of your post.
 - [x] Checked that the build passes?
 - [x] Checked your spelling (you can use `npm install` followed by `npx mdspell "**/{FILE_NAME}.md" --en-gb -a -n -x -t` if that's your thing)
 - [x] Ensured that your author profile contains a profile image, and a brief description of yourself? (make it more interesting than just your job title!)
 - [x] Optimised any images in your post? They should be less than 100KBytes as a general guide.

Posts are reviewed / approved by your Regional Tech Lead.

### CODE REVIEW

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+date: 2024-10-28 12:00:00 Z
+categories:
+- Testing
+tags:
+- Testing
+- Automation
+- API
+- Open Source
+- Karate
+summary: In this blog i'll introduce the Karate Test Automation Framework and talk about some of the fun and
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: I'll

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  interesting features it provides.   
+author: sdewar
+---
+
+# Introduction
+Karate is an automation framework designed to make automation testing easy, super readable and more reliable than other 
+offerings in the open source space - don't even ask me how many times I've been bitten by Selenium's reluctance to play 
+nice with UI elements or been snowed under trying to get my head around a complex multi-util, multi-file test scenario.
+
+Karate is essentially a Gherkin-like programming language, with the ability to use Java & JavaScript almost seamlessly
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: Might be worth a link to Gherkin for those unfamiliar

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+in order to handle more complex and unique functionality should you require it. It's primarily an API Automation
+tool, but it also handily also provides very good UI Automation capabilities, Load Testing, Kafka Testing & Image 
+Comparison Testing, amongst other features. 
+
+# So Stuart, what are you going to show us?
+There's a whole host of interesting & handy features of the Framework, but I'll focus on a few that make API testing
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: ampersand feels unnecessary? Also not sure framework needs a capital

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+easy, efficient and clear.
+
+## Test Readability
+I'll preface this with the fact Cucumber/BBD is more readable - but that requires a huge amount of effort in 
+implementing the step-definitions correctly, which means higher time investment and ultimately less readable code in the
+background. 
+
+Karate lets us use keywords straight out the box, no need for step-definitions & no need for code, everything is
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: & again?

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+Scenario: Given we send a request to the 'users' endpoint, a 200 response is returned
+
+  Given path 'users'
+  When method GET
+  Then status 200
+  And match response.data[*].first_name contains "Tobias"
+~~~
+
+That's all we need - this test that will verify that the `users` endpoint returns a 200. Additionally, we run an
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: "this test that will verify that the" 😕

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown (hunk 4 above)
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: If you're not familiar with step definitions or Karate keywords I'm not sure it's super clear what your point is here. Might be worth highlighting by referencing the example below - e.g. "For instance in the example below we use the keywords foo and bar to do blah without the need for any setup"

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+I'll preface this with the fact Cucumber/BBD is more readable - but that requires a huge amount of effort in 
+implementing the step-definitions correctly, which means higher time investment and ultimately less readable code in the
+background. 
+
+Karate lets us use keywords straight out the box, no need for step-definitions & no need for code, everything is
+already there hidden away, we really don't need to dive any deeper into the technical behind-the-scenes implementation 
+in order to get testing our services.
+
+Let's take a look - say we want to send a request to an endpoint to list users (we'll use [reqres](https://reqres.in/) 
+for the examples in this Blog):
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: in this Blog -> in this post

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown (hunk 5 above)
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: minor but might be worth saying "a http `200` response code" just to be clearer

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+Karate is an automation framework designed to make automation testing easy, super readable and more reliable than other 
+offerings in the open source space - don't even ask me how many times I've been bitten by Selenium's reluctance to play 
+nice with UI elements or been snowed under trying to get my head around a complex multi-util, multi-file test scenario.
+
+Karate is essentially a Gherkin-like programming language, with the ability to use Java & JavaScript almost seamlessly
+in order to handle more complex and unique functionality should you require it. It's primarily an API Automation
+tool, but it also handily also provides very good UI Automation capabilities, Load Testing, Kafka Testing & Image 
+Comparison Testing, amongst other features. 
+
+# So Stuart, what are you going to show us?
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: Might just  be me but the format of "What are you going to show us?" as a H1 and then the main meat of the post nested underneath feels a bit odd - I'd be tempted to ditch that heading and lift everything else up to the same level

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+The main point here though is just showcasing how easy it is to write our test scenarios and how readable they can be, 
+meaning less time is spent understanding the scenarios and more time finding bugs!
+
+We don't even have to stick to Given, When, Then syntax. It is completely interchangeable - the Karate way is to write
+your Scenarios in the most readable way that makes sense to you and your test flows.
+
+## Feature File Re-use, Call Single & Caching
+This next feature is both a nice saver in terms of execution cost/time but also in terms of code re-usability. I know 
+I'll mention that this is necessarily not a great pattern in terms of test readability, but there certainly still
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: "I know I'll mention" feels a weird way to put it - maybe just "It's true that this is not necessarily a great pattern in terms of test readabilit..."

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+is a time and place for it.
+
+Say we want to add some authentication to our request, we typically will need to get a token in which we can 
+authenticate with, save the token from the response and then add our Authorization header like so:
+
+~~~karate
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: had to read this a couple of times to get what was going on. Might be worth a quick rundown on the main bits rather than relying on the comments. Same with the above example - a quick lowdown on the key bits (Feature, Background, *) could be helpful

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+  Given path 'users'
+  And header Authorization = `Bearer ${token}`
+  When method GET
+  Then status 200
+~~~
+
+Obviously once our Test Suite grows arms and legs, we don't want to authenticate for every single request, especially
+for large complex test flows that are happy with the same authorization token - enter `callSingle`. 
+
+Karate allows us to move any piece of commonly used functionality out into a Feature File that can be "called" (via
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: not sure called needs quotes?

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+  When method GET
+  Then status 200
+~~~
+
+Obviously once our Test Suite grows arms and legs, we don't want to authenticate for every single request, especially
+for large complex test flows that are happy with the same authorization token - enter `callSingle`. 
+
+Karate allows us to move any piece of commonly used functionality out into a Feature File that can be "called" (via 
+`call`, `call once` or `callSingle`). Here, we can move the authentication request into a `callSingle` call and put it
+into the Background. Usually, all code within the `Background` is executed for every Scenario in that Feature File, but
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: Background seems to have both backticks and not here

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+inside the Feature file. `call` is especially handy when you have a complex flow of requests and test steps that "muddy"
+the waters of a Scenario that you can just move out into another file and replace with a one-liner.
+
+Within the world of development, we try to re-use code as much as possible - but within test automation this can
+seriously hamper test readability and time spent understanding, debugging & fixing your tests. Tests should be clear
+on what they are testing and if that means we have to re-use code then I think that's absolutely fine. In my experience 
+of using Karate, I feel that it provides a really nice middle-ground that empowers the tester to make the decision on 
+how readable their test scenarios are.
+
+## Hybrid Scenarios (+ a sneak peek into Karate UI Automation)
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: + -> plus?

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+  # This doesn't actually exist, but used just as an example
+  Given driver `${baseUrl}/login`
+  And waitFor(usernameLocator).input('sdewar@scottlogic.com')
+  And waitFor(passwordLocator).input('test123')
+  When waitFor(signInButtonLocator).click()
+  Then waitForUrl(`${baseUrl}/home`)
+~~~
+
+Firstly, we have our API call to create the new user. We then need to tell Karate that our `baseUrl` has now changed to
+the UI one. Then we instantiate a `driver` instance, and use locators stored in a `homePagelocators.json` file in order
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: might be worth linking to the docs a bit for some things like `driver` for example https://karatelabs.github.io/karate/karate-core/#driver

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+tell Karate to wait until they are rendered on screen. We can also use chaining to perform actions as soon as the 
+element is found - removing the need for multiple lines of code or storing the element into a variable.
+
+## Dynamic Scenario Outlines
+Lastly, I want to talk about Dynamic Scenario Outlines. 
+
+If you're used to Cucumber then you've probably got an understanding of Scenario Outlines - they let you run through
+the exact same test steps but with your variables & data driven directly from a table, for example:
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: & 😄

Comment (dogle-scottlogic) [reply _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: I'll not comment each one going forward...

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+            "last_name": "Ferguson",
+            "avatar": "https://reqres.in/img/faces/8-image.jpg"
+        }
+    ]
+}
+~~~
+
+This is great for running through a bulk set of tests where we don't necessarily know some vital inputs before 
+execution. There's a few others ways to do something similar in Karate, but another benefit of Dynamic Scenario Outlines
+is that Karate will still respect any parallel execution configuration, meaning we can run this across `x` threads with
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: I think n is probably a more common variable than x here - having x in backticks made me think I'd missed a reference to it somewhere above. Could you replace `x` with maybe "multiple"?

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+no extra config outside of the parallel runner configuration.
+
+# Conclusion
+The aim of this blog post was just to show some features of the Karate Test Framework that make API (and UI!) testing
```
Comment (dogle-scottlogic) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: The conclusion falls a bit flat for me personally - would be good to reword this as a summary of your thoughts on things here. 
I'd say something like "look out for future posts where I'll dive deeper into..."
I'd remove the line and resources heading and just make that paragraph part of the conclusion text personally.

**Review by dogle-scottlogic — COMMENTED** [review]
Great stuff Stu! Couple of thoughts in general,
- Could do with a scattering of links to some things folks might want to read more about / might not have heard of (Cucumber?)
- There's a lot of abbreviation which feels a bit odd, there's no character limit (e.g. & and +)
- Could do with a little more hand holding through the examples personally to make it easier to understand what's happening for the layman 

Give me a shout if you want to chat on anything 😄

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown (hunk 7 above)
Comment (chriswilty) [reply _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: Correct. We shouldn't use any h1 in our posts, lowest should be h2, because the post title is the h1.

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown (hunk 9 above)
Comment (chriswilty) [reply _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: "we will need to request a token, and then add that as a header in all subsequent requests" ??

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+
+~~~karate
+Feature: Delete Users using Dynamic Scenario Outline
+
+Background:
+
+  # Tell Karate that this is our "base url" for all Scenarios in this Feature File
+  * baseUrl 'https://reqres.in/api'
+    
+@setup
```
Comment (chriswilty) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: This annotation is used for the first time here, I missed it first time through... Can you call it out in the description?

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+  Given path `users/${id}`
+  When method DELETE
+  Then status 204
+
+Examples: 
+  | karate.setup().userData |
+~~~
+
+Since we told our Scenario Outline to use the `userData` array as its data source via our `Examples` table, it will have access 
+to any key-value pair present in that array. Here's a snippet of the data array from the `users` endpoint:
```
Comment (chriswilty) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: An array doesn't have key-value pairs, it's the user entries that have those. Are you saying that we can use the `id` key  which will be replaced with the value? That was a little unclear to me.

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+    ]
+}
+~~~
+
+This is great for running through a bulk set of tests where we don't necessarily know some vital inputs before 
+execution. There's a few others ways to do something similar in Karate, but another benefit of Dynamic Scenario Outlines
+is that Karate will still respect any parallel execution configuration, meaning we can run this across `x` threads with
+no extra config outside of the parallel runner configuration.
+
+# Conclusion
```
Comment (chriswilty) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: Level 2 heading here:

```
## Conclusion
```

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+- Automation
+- API
+- Open Source
+- Karate
+summary: In this blog i'll introduce the Karate Test Automation Framework and talk about some of the fun and
+  interesting features it provides.   
+author: sdewar
+---
+
+# Introduction
```
Comment (chriswilty) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: Change this to level2 heading: `## Introduction`

In fact, you arguably don't need "Introduction" as a heading at all, because we can expect some introduction at the top of a blog post.

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown (hunk 16 above)
Comment (chriswilty) [reply _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: What Dave said: don't be shy!

Tell us there'll be a follow-up, and that might even _encourage_ you to write it 😉

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+
+## Resources
+
+The Documentation for Karate is really solid, with lots of examples and good explanations of each bit of functionality.
+There's also a very active community on Stack Overflow asking and answering questions - you'll find the author of 
+Karate answering a lot of queries.
+
+[Karate Documentation](https://karatelabs.github.io/karate/)
+
+[Stack Overflow](https://stackoverflow.com/questions/tagged/karate)
```
Comment (chriswilty) [inline _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: Link text should be more descriptive, e.g. "Karate on Stack Overflow"

**Review by chriswilty — COMMENTED** [review]
Nice read 👏 

Not much else to add beyond what Dave has already said. I know this is not easy, but try to imagine you are reading those code blocks for the first time: are all the important parts indicated?

File: _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown (hunk 7 above)
Comment (dogle-scottlogic) [reply _posts/2024-10-28-an-introduction-into-karate-test-automation.markdown]: I did wonder about that - we seem a bit inconsistent in past posts - maybe one for the checklist on PRs

File: _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown [hunk 22]
```diff
@@ -0,0 +1,379 @@
...
+easy, efficient and clear.
+
+## Test Readability
+I'll preface this with the fact [Cucumber](https://cucumber.io/docs/cucumber/) is more readable - but that requires a 
+huge amount of effort in implementing the step-definitions correctly, which means higher time investment and ultimately 
+less readable code in the background. 
+
+Karate lets us write our test scenarios straight out the box - there's no need to write Java glue code or 
+step-definitions like Cucumber. The example below showcases this perfectly - there's no additional setup needed
+or extra digging into the technical behind-the-scenes implementation in order to get testing our services.
```
Comment (dogle-scottlogic) [inline _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown]: the technical, behind-the-scenes implementation

File: _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown [hunk 23]
```diff
@@ -0,0 +1,379 @@
...
+our `baseUrl` has now changed to the UI because when we instantiate a 
+[driver](https://karatelabs.github.io/karate/karate-core/#driver) instance, it'll automatically navigate to the 
+currently configured `baseUrl`. 
+
+Karate's UI test features are really robust in the sense of avoiding flakey tests and waiting on elements appearing on 
+screen. One example is `waitFor(locator)` - which will wait until a given locator is present on screen. We can also 
+chain commands on the back of the locator being found as opposed to waiting for the locator, storing it, then performing
+an action. 
+
+This means we have a tidy of steps that will spin up our driver, wait for elements to appear, perform
```
Comment (dogle-scottlogic) [inline _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown]: "This means we have a tidy of steps"?

File: _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown [hunk 24]
```diff
@@ -0,0 +1,379 @@
...
+        }
+    ]
+}
+~~~
+
+Let's say we wanted to clean up our test environment of users, but in order to do that we need to know all user 
+ID's and then send a DELETE request to `https://reqres.in/api/users/${id}` for each one.
+
+In the below example, we have a `@setup` Scenario which is our data setup for our Scenario Outline.
+Up until now we are used to the `Background` section running for every Scenario, but that's not the case here, so we need
```
Comment (dogle-scottlogic) [inline _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown]: might just be me - why is it not the case? Is it because @setup causes that to behave differently? If so why do we need to define the baseUrl in background at all here?

File: _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown [hunk 25]
```diff
@@ -0,0 +1,379 @@
...
+I hope you've enjoyed some brief examples of a few features of the Karate Test Automation framework that make API
+(and UI!) testing really easy, readable and accessible. Testing complicated back-end services, which typically are not 
+very human-friendly without an interactive UI, doesn't need to be complicated - Karate gives us scope to translate
+this layer of testing into clean and concise test scenarios that hopefully improves the test experience for everyone.
+
+We have only scratched the surface in terms of everything that the framework offers and in a subsequent blog we can take
+a deeper dive into some of the other key features such as exploring the comprehensive assertion features, image 
+comparison, built in HTML reporting or data driven tests via tags.
+
+The Documentation for Karate is really solid, with lots of examples and good explanations of each bit of functionality.
```
Comment (dogle-scottlogic) [inline _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown]: capital D not needed?

**Review by dogle-scottlogic — COMMENTED** [review]
Loads better 👍  I was able to follow the examples a lot easier this time round. Minor one but I can't access the deployed page as the time is set to 12:00:00 Z

File: _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown (hunk 24 above)
Comment (sdewar-scottlogic) [reply _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown]: Yeah its a weird quirk of the implementation that tripped me up for many hours, @setup runs before background so @setup doesn't know about anything we've set up in background.

We still need it in the background so our Scenario Outline (and any other scenarios in the file) know what the baseUrl is. I'll update and perhaps link to that specific part of the docs here.

Comment (dogle-scottlogic) [reply _posts/2024-11-01-an-introduction-into-karate-test-automation.markdown]: ah ok, at least it's not just me being dumb 😆

**Review by dogle-scottlogic — APPROVED** [review]
lgtm 🚀

**Review by chriswilty — APPROVED** [review]
Nice 👌
